prov/efa/src/efa_mr.c

```diff
@@ -235,102 +228,157 @@ struct fi_ops efa_mr_ops = {
-	int ret;
+	int ret = 0;
+
+	/* To support Emulated RMA path, if the access is not supported
+	 * by EFA, modify it to FI_SEND | FI_RECV
+	 */
+	core_access = mr_attr->access;
+	if (!core_access || (core_access & ~EFA_MR_SUPPORTED_PERMISSIONS))
+		core_access = FI_SEND | FI_RECV;
+
+	/* Local read access to an MR is enabled by default in verbs */
+	if (core_access & FI_RECV)
+		fi_ibv_access |= IBV_ACCESS_LOCAL_WRITE;
+
+	if (efa_mr->domain->ctx->device_caps & EFADV_DEVICE_ATTR_CAPS_RDMA_READ)
+		fi_ibv_access |= IBV_ACCESS_REMOTE_READ;
+
+	efa_mr->ibv_mr = ibv_reg_mr(efa_mr->domain->ibv_pd, 
+				    (void *)mr_attr->mr_iov->iov_base,
+				    mr_attr->mr_iov->iov_len, fi_ibv_access);
+	if (!efa_mr->ibv_mr) {
+		EFA_WARN(FI_LOG_MR, "Unable to register MR: %s\n",
+				fi_strerror(-errno));
+		return -errno;
+	}
+
+	efa_mr->mr_fid.mem_desc = efa_mr->ibv_mr;
+	efa_mr->mr_fid.key = efa_mr->ibv_mr->rkey;
+	efa_mr->peer.iface = mr_attr->iface;
+	if (mr_attr->iface == FI_HMEM_CUDA)
+		efa_mr->peer.device.cuda = mr_attr->device.cuda;
+	assert(efa_mr->mr_fid.key != FI_KEY_NOTAVAIL);
+
+	mr_attr->requested_key = efa_mr->mr_fid.key;
+
+	ret = ofi_mr_map_insert(&efa_mr->domain->util_domain.mr_map, attr,
+				&efa_mr->mr_fid.key, &efa_mr->mr_fid);
+	if (ret) {
+		EFA_WARN(FI_LOG_MR,
+			"Unable to add MR to map buf (%s): %p len: %zu\n",
+			fi_strerror(-ret), mr_attr->mr_iov->iov_base,
+			mr_attr->mr_iov->iov_len);
+		return ret;
+	}
+	if (efa_mr->domain->shm_domain && rxr_env.enable_shm_transfer) {
+		ret = fi_mr_regattr(efa_mr->domain->shm_domain, attr,
+				    flags, &efa_mr->shm_mr);
+		if (ret) {
+			EFA_WARN(FI_LOG_MR,
+				"Unable to register shm MR buf (%s): %p len: %zu\n",
+				fi_strerror(-ret), mr_attr->mr_iov->iov_base,
+				mr_attr->mr_iov->iov_len);
+			fi_close(&efa_mr->mr_fid.fid);
+			ofi_mr_map_remove(&efa_mr->domain->util_domain.mr_map,
+						efa_mr->mr_fid.key);
+			return ret;
+		}
+	}
+	return 0;
+}
+
+static int efa_mr_regattr(struct fid *fid, const struct fi_mr_attr *attr,
+			  uint64_t flags, struct fid_mr **mr_fid)
+{
+	struct fid_domain *domain_fid;
+	struct efa_mr *efa_mr = NULL;
+	int ret = 0;
 
 	if (flags && flags != OFI_MR_NOCACHE) {
-		EFA_WARN(FI_LOG_MR, "Unsupported flag type. requested[0x%" PRIx64 "] supported[0x%" PRIx64 "]\n",
-				flags, (uint64_t) OFI_MR_NOCACHE);
-		ret = -FI_EBADFLAGS;
-		goto err;
+		EFA_WARN(FI_LOG_MR, "Unsupported flag type. requested"
+			 "[0x%" PRIx64 "] supported[0x%" PRIx64 "]\n",
+			 flags, (uint64_t) OFI_MR_NOCACHE);
+		return -FI_EBADFLAGS;
 	}
 
 	if (fid->fclass != FI_CLASS_DOMAIN) {
-		EFA_WARN(FI_LOG_MR,
-			 "Unsupported domain. requested[0x%" PRIx64 "] supported[0x%" PRIx64 "]\n",
+		EFA_WARN(FI_LOG_MR, "Unsupported domain. requested"
+			 "[0x%" PRIx64 "] supported[0x%" PRIx64 "]\n",
 			 fid->fclass, (uint64_t) FI_CLASS_DOMAIN);
-                ret = -FI_EINVAL;
-                goto err;
-        }
+		return -FI_EINVAL;
+	}
 
-	if (access & ~EFA_MR_SUPPORTED_PERMISSIONS) {
```

Comment:
Is this check still not valid?

Reply:
The only reason for this check in efa layer was for setting permission for fi_ibv_access for ibv_reg_mr call. rxr layer would always set access to FI_SEND | FI_RECV when calling fi_mr_reg on efa_mr. So, even if the app provides and access which is not supported by efa, rxr layer would send FI_SEND | FI_RECV as access to efa layer. 
With this PR we are doing the same emulation in efa_mr_reg_impl. I can add the check back if we plan on removing emulation.